# Regression Metrics

#### We measure how far predictions are from actual values.

In [2]:
import pandas as pd

In [11]:
y = [10, 20, 30, 40]
y_hat = [12,18,25,42]
df = pd.DataFrame({'y' : y, 
                   'y_hat' : y_hat})
df.head()

,y,y_hat
0,10,12
1,20,18
2,30,25
3,40,42


## MAE - Mean Absolute Error
- On Average, how much was the model wrong 

In [13]:
df['mod_y_y_hat'] = abs(df['y'] - df['y_hat'])
df.head()

,y,y_hat,mod_y_y_hat
0,10,12,2
1,20,18,2
2,30,25,5
3,40,42,2


In [17]:
mae = (df['mod_y_y_hat'].sum())/(len(df))
print(f'prediction is approximately actual ± {mae}')

prediction is approximately actual ± 2.75


MAE means Mean Absolute Error — it ignores the direction of the error.

If MAE = 2.75, it means:

✔ Sometimes your model predicts 2.75 higher  
✔ Sometimes your model predicts 2.75 lower  
✔ The average mistake size is 2.75

In math form:

MAE = 2.75  ⇒  Prediction error ≈ ±2.75

So your model is off by about 2.75 units on average.


## MSE — Mean Squared Error

$$
\text{MSE} = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2
$$


In [18]:
df.head()

,y,y_hat,mod_y_y_hat
0,10,12,2
1,20,18,2
2,30,25,5
3,40,42,2


In [23]:
mse = sum([i*i for i in df['mod_y_y_hat']])/len(df)
mse 

9.25

What does it mean when MSE is high 


1. Your model is too simple → predictions are far from actual.

In [36]:
print('Lets just say')
print(f'If my actual y is {[i for i in df['y']]} and my model predictions are {[25 for i in range(4)]}')
print(f'sum of error is {sum([(i - 25)**2 for i in df['y']])} and the mse is {sum([(i - 25)**2 for i in df['y']])/len(df['y'])}')
print('This is the case of High MSE because MODEL IS WRONG (Underfitting)')

Lets just say
If my actual y is [10, 20, 30, 40] and my model predictions are [25, 25, 25, 25]
sum of error is 500 and the mse is 125.0
This is the case of High MSE because MODEL IS WRONG (Underfitting)


2. Model is good, but one bad data point destroys MSE.

In [41]:
y = [10, 20, 30, 200]
y_pred = [12, 22, 28 ,50]
error = [a - b for a,b in zip(y, y_pred)]
mse = sum([i*i for i in error])/len(y)
print(f'NOT model’s fault — last point is too far.One outlier added 22500 to MSE which is ------>.{mse}')



NOT model’s fault — last point is too far.One outlier added 22500 to MSE which is ------>.5628.0


| Cause            | Why MSE is high?                     | Blame                  |
|------------------|----------------------------------------|------------------------|
| Underfitting     | Model is too simple                    | Model                  |
| Outliers         | One big error gets squared (explodes)  | Data                   |
| Missing features | Model is blind to important patterns   | Inputs / Preprocessing |


## RMSE — Root Mean Squared Error
# Root Mean Squared Error (RMSE)



$$
\text{RMSE} = \sqrt{\text{MSE}}
$$

Expanded form:

$$
\text{RMSE} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$


In [48]:
rmse = math.sqrt(mse)
rmse


75.0199973340442

#### Why RMSE Is Important (Concise Points)

1. Same units as target
- MSE is in squared units (₹², kg², etc.)
- RMSE = sqrt(MSE) → converts error back to original units
- Example: RMSE = 3000 → model is off by ~₹3000

2. Easy to judge error size
- Compare RMSE with typical value of y
- Small RMSE vs y → good model
- Large RMSE vs y → bad model

3. Punishes big errors strongly
- Comes from squaring errors
- One large mistake increases RMSE a lot
- Helps detect outliers and unstable predictions

4. Standard metric everywhere
- Used in Kaggle, industry ML models, research
- More intuitive than MSE

5. Easy to compare models
- Smaller RMSE = better model
- Example:
  Model A: 13
  Model B: 18
  Model C: 10
  → C is best

6. RMSE vs MAE gives insights
- RMSE >> MAE → big errors/outliers exist
- RMSE ≈ MAE → errors are uniform

7. Helps detect overfitting
- Train RMSE << Test RMSE → overfitting
- RMSE clearly shows the gap


75.0199973340442

#### MAE vs MSE vs RMSE — Side-by-Side Comparison

| Metric | Meaning                 | Punishes Big Errors? | Easy to Understand? | Sensitive to Outliers? |
|--------|--------------------------|------------------------|-----------------------|--------------------------|
| MAE    | Average absolute error   | ❌ No                  | ✔ Very easy           | Low                      |
| MSE    | Average squared error    | ✔ Yes (squares)       | ❌ Hard (units squared)| High                     |
| RMSE   | Square root of MSE       | ✔ Yes (less than MSE) | ✔ Easy (same units)   | Medium                   |


# Classification Metrics

In [4]:
data = {
    "Customer": [1,2,3,4,5,6,7,8,9,10],
    "Income": ["50K","30K","45K","25K","70K","20K","80K","22K","40K","28K"],
    "Credit Score": [720,600,650,550,780,500,800,580,670,560],
    "Actual (Default)": [0,1,0,1,0,1,0,1,0,1],
    "Predicted": [0,1,1,1,0,0,0,1,0,0]
}

df = pd.DataFrame(data)
df


,Customer,Income,Credit Score,Actual (Default),Predicted
0,1,50K,720,0,0
1,2,30K,600,1,1
2,3,45K,650,0,1
3,4,25K,550,1,1
4,5,70K,780,0,0
5,6,20K,500,1,0
6,7,80K,800,0,0
7,8,22K,580,1,1
8,9,40K,670,0,0
9,10,28K,560,1,0


#### COVID Test Example — Understanding TP, TN, FP, FN

A hospital uses a test to predict if a person has COVID.

- Actual = real truth
- Predicted = test result

Below are the 4 possible outcomes:

------------------------------------------------------------

1. TRUE POSITIVE (TP)
- Predicted: Positive
- Actual: Positive
- Meaning: Correct prediction
- Interpretation: “Test says you have COVID, and you really have COVID.”

Example:
Actual = Positive
Predicted = Positive
→ TP

------------------------------------------------------------

2. TRUE NEGATIVE (TN)
- Predicted: Negative
- Actual: Negative
- Meaning: Correct prediction
- Interpretation: “Test says you don’t have COVID, and you really don’t.”

Example:
Actual = Negative
Predicted = Negative
→ TN

------------------------------------------------------------

3. FALSE POSITIVE (FP)
- Predicted: Positive
- Actual: Negative
- Meaning: Wrong prediction
- Interpretation: “Test says you have COVID, but you actually DON’T.”
- Risk: Causes panic or unnecessary treatment.
- Also called: Type-1 Error

Example:
Actual = Negative
Predicted = Positive
→ FP

------------------------------------------------------------

4. FALSE NEGATIVE (FN)
- Predicted: Negative
- Actual: Positive
- Meaning: Wrong and dangerous prediction
- Interpretation: “Test says you are fine, but you actually HAVE COVID.”
- Risk: Person may spread the disease.
- Also called: Type-2 Error

Example:
Actual = Positive
Predicted = Negative
→ FN


In [7]:
df

,Customer,Income,Credit Score,Actual (Default),Predicted
0,1,50K,720,0,0
1,2,30K,600,1,1
2,3,45K,650,0,1
3,4,25K,550,1,1
4,5,70K,780,0,0
5,6,20K,500,1,0
6,7,80K,800,0,0
7,8,22K,580,1,1
8,9,40K,670,0,0
9,10,28K,560,1,0


#### TP, TN, FP, FN from the Dataset

True Positives (TP)
Actual = 1, Predicted = 1
Customers: 2, 4, 8
TP = 3

True Negatives (TN)
Actual = 0, Predicted = 0
Customers: 1, 5, 7, 9
TN = 4

False Positives (FP)
Actual = 0, Predicted = 1
Customers: 3
FP = 1

False Negatives (FN)
Actual = 1, Predicted = 0
Customers: 6, 10
FN = 2


#### Confusion Matrix

| Actual \ Predicted | Default (1) | No Default (0) |
|--------------------|-------------|----------------|
| Default (1)        | TP = 3      | FN = 2         |
| No Default (0)     | FP = 1      | TN = 4         |
